dart/lib/src/reference.dart

```diff
@@ -0,0 +1,416 @@
+      return _readUInt(_offset, _parentWidth);
+    }
+    if (_valueType == ValueType.IndirectInt) {
+      return _readInt(_indirect, BitWidthUtil.fromByteWidth(_byteWidth));
+    }
+    if (_valueType == ValueType.IndirectUInt) {
+      return _readUInt(_indirect, BitWidthUtil.fromByteWidth(_byteWidth));
+    }
+    return null;
+  }
+
+  /// Returns [double] value or null otherwise.
+  double get doubleValue {
+    if (_valueType == ValueType.Float) {
+      return _readFloat(_offset, _parentWidth);
+    }
+    if (_valueType == ValueType.IndirectFloat) {
+      return _readFloat(_indirect, BitWidthUtil.fromByteWidth(_byteWidth));
+    }
+    return null;
+  }
+
+  /// Returns [num] value or null otherwise.
+  num get numValue {
+    if (isDouble) {
+      return doubleValue;
+    }
+    return intValue;
+  }
+
+  /// Returns [String] value or null otherwise.
+  String get stringValue {
+    if (_valueType == ValueType.String || _valueType == ValueType.Key) {
+      return utf8.decode(_buffer.buffer.asUint8List(_indirect, length));
+    }
+    return null;
+  }
+
+  /// Returns [Uint8List] value or null otherwise.
+  Uint8List get blobValue {
+    if (_valueType == ValueType.Blob) {
+      return _buffer.buffer.asUint8List(_indirect, length);
+    }
+    return null;
+  }
+
+  /// Can be used with an [int] or a [String] value for key.
+  /// If the underlying value in FlexBuffer is a vector, then use [int] for access.
+  /// If the underlying value in FlexBuffer is a map, then use [String] for access.
+  /// Returns [Reference] value or null. Does not throw out of bounds exception.
+  Reference operator [](Object key) {
+    if (key is int && ValueTypeUtils.isAVector(_valueType)) {
+      var index = key;
+      if(index >= length) {
+        return null;
+      }
+      var elementOffset = _indirect + index * _byteWidth;
+      var flx = Reference._(_buffer, elementOffset, BitWidthUtil.fromByteWidth(_byteWidth), 0);
+      flx._byteWidth = 1;
+      if (ValueTypeUtils.isTypedVector(_valueType)) {
+        flx._valueType = ValueTypeUtils.typedVectorElementType(_valueType);
+        return flx;
+      }
+      if(ValueTypeUtils.isFixedTypedVector(_valueType)) {
+        flx._valueType = ValueTypeUtils.fixedTypedVectorElementType(_valueType);
+        return flx;
+      }
+      var packedType = _buffer.getUint8(_indirect + length * _byteWidth + index);
+      return Reference._(_buffer, elementOffset, BitWidthUtil.fromByteWidth(_byteWidth), packedType);
+    }
+    if (key is String && _valueType == ValueType.Map) {
+      var index = _keyIndex(key);
+      if (index != null) {
+        return _valueForIndex(index);
+      }
+    }
+    return null;
+  }
+
+  /// Get an iterable if the underlying flexBuffer value is a vector.
+  /// Otherwise throws an exception.
+  Iterable<Reference> get vectorIterable {
+    if(isVector == false) {
+      throw Exception('Value is not a vector. It is: ${_valueType}');
```

Comment:
`StateError`, or perhaps a custom exceptoin.

Also, on the line above, missing a space after `if`.